### PR TITLE
Tr/bit trie

### PR DESCRIPTION
**Pull request by travitch**
This PR is a draft of a bit-level trie to replace the byte trie (radix tree) used currently.  It significantly reduces compile times and memory use during compilation.  However, it doesn't work yet.

### CODE REVIEW

**Comment by travitch** [comment]
Okay, this seems to be in pretty good shape at this point. The open issue is that many tests fail because the ByteTrie hasn't been updated.  Should we update it?  It isn't used and there isn't really a reason to keep it.

File: dismantle-tablegen/src/Dismantle/Tablegen/TH.hs [hunk 1]
```diff
@@ -78,10 +82,14 @@ loadISA isa path overridePaths = do
 genISA :: ISA -> FilePath -> [FilePath] -> DecsQ
 genISA isa path overridePaths = do
   desc <- loadISA isa path overridePaths
-  genISADesc isa desc [path]
+  genISADesc isa desc [path] Nothing
 
-genISADesc :: ISA -> ISADescriptor -> [FilePath] -> DecsQ
-genISADesc isa desc paths = do
+-- | This function creates all of the definitions for an ISA dissasembler/assembler
```
Comment (kquick) [inline dismantle-tablegen/src/Dismantle/Tablegen/TH.hs:87]: disassembler

File: dismantle-arm-xml/README.org [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+The major differences compared to the rest of dismantle are:
+- The data source is from ARM instead of LLVM
+- It uses pre-computed parse tables that are checked in (in binary form)
+
+* Pre-computed Parse Tables
+
+Using our current parse table construction algorithm, constructing the parse tables for the Thumb encoding takes over an hour.  This makes building the disassembler repeatedly untenable.  To mitigate this, there is a tool, dismantle-arm-xml-tablegen, which can be used to pre-compute the necessary parse tables and persist them in a serialized binary format.  During compilation, the Template Haskell in this package loads the pre-computed tables.
+
+That tool should be run from the root of this package (not the repository root).  It will save the necessary pre-computed tables in the data directory of this package.  The tool supports a ~--parallel~ flag to compute all of the necessary table variants in parallel.
```
Comment (kquick) [inline dismantle-arm-xml/README.org:13]: Should probably add a note about the amount of memory needed to run the tool (directly or with --parallel).  And also the amount of memory needed to compile this package using pre-built tables.

File: dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+         let pats = map fst (M.toList matchingPatterns)
+         let mnemonics = catMaybes (flip M.lookup mapping <$> pats)
+
+         error ("Overlapping bit pattern " ++ show (bitIndex, bitsSoFar, mapping, pats, mnemonics))
+  where
+    -- Now find the length of the smallest matching pattern
+    --
+    -- If there are multiple matches and one is shorter than the rest, we can
+    -- use this to say that the unique shortest pattern supersedes the others.
+    -- minPatLen = minimum (patternBits <$> M.keys matchingNegativePatterns)
```
Comment (kquick) [inline dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs:210]: Remove this?

File: dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+              --
+              --                 abc
+              -- Bit String: ????????
+              -- Mask:       00001100
+              --
+              -- - In @a@, length(bs)=5, highestBit(mask)=5
+              -- - In @b@, length(bs)=6, highestBit(mask)=5
+              -- - In @c@, length(bs)=7, highestBit(mask)=5
+              --
+              -- So we cannot apply the mask if until
```
Comment (kquick) [inline dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs:304]: "if until" -> "until:"

File: dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+    -- Return 'True' if @target@ is more specific than @p@.
+    isMoreSpecific target (p, _) = requiredBitCount target > requiredBitCount p
+    requiredBitCount bs = sum [ popCount w
+                              | w <- BS.unpack (DTP.requiredMask bs)
+                              ]
+
+-- | Convert from indexes in linked tables to indexes suitable for the flat
+-- tables we are constructing for the ultimate BitTrie
+--
+-- Unlike the 'ByteTrie' (which has 256 entries per table), this one only has 2.
```
Comment (kquick) [inline dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs:340]: Remove: there is no ByteTrie anymore.

File: dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+linkedToFlatIndex :: DTP.LinkedTableIndex -> DTP.FlatTableIndex
+linkedToFlatIndex (DTP.LTI i)
+  | i < 0 = DTP.FTI i
+  | otherwise = DTP.FTI (i * 2)
+
+-- | A helper type to track accumulated concrete bits
+--
+-- The ByteTrie uses a ByteString since that is fairly natural.  We could do the
```
Comment (kquick) [inline dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs:348]: Ditto (or reword).

File: dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+
+Observation: construction is DFS.  In a bit pattern like
+
+> P = |1|0|?|?|1|?|?|1|
+>      0 1 2 3 4 5 6 7
+
+where the second operand is not allowed to be 0b11.  We traverse out to the end
+of the bit-string depth first covering suffixes (starting from bit 5):
+
+- 0b001   # operand is 0b00 (and fine)
```
Comment (kquick) [inline dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs:485]: This is still confusing (re my previous note).  Is `0b001` indicating bit 7 is set and bits 5 and 6 are zero?  Or is bit 5 set and bits 6 and 7 zero?  Or are these bits 0-2?

File: dismantle-arm-xml/README.org (hunk 2 above)
Comment (travitch) [reply dismantle-arm-xml/README.org:13]: It is negligible - about 4-6GB now

**Comment by travitch** [comment]
Merged manually.

**Pull request closed without merging**